Emoji syntax

calculuschild opened the pull request:
Following the recent addition of multiple icon fonts, this adds a proper emoji syntax `:emoji:` like Github. Supports any custom emojis we want; just have to list them out in a .js file. Also, an autocomplete dropdown will appear showing all matching icons!

I have done this so far with only dicefont, but it is easy to add the rest, if a little tedious. Manually added `fas-heart` and `fas-star` just to make sure it works. The list uses this format:

```
{
displayName: CSSClasses,
displayName: CSSClasses,
}
```

`displayName` is both what will appear in the autosuggest, and in the `:displayName:`
![image](https://github.com/naturalcrit/homebrewery/assets/5878534/28da1b3e-11c7-404d-b48b-ad6b42af00d3)

![image](https://github.com/naturalcrit/homebrewery/assets/5878534/de0c5b0c-abf8-4098-bc5b-149895edb9e3)

Again, taking inspiration from Obsidian, we might choose CamelCase for our display names, and choose names that are more likely searched, such as including "dice" in the name like **DfDiceD10-1** instead of **df-d10-1** for example.

![image](https://github.com/naturalcrit/homebrewery/assets/5878534/ead94e5d-c604-47c6-ae27-accc271101e3)

Gazook89 commented:
without altering the font, is there any method of using an alternate display name?  So that it could just be `:d10-2:`?

calculuschild commented:
They can be whatever display name we want as long as there are no conflicts. However I think we should keep some identifier to differentiate between sets. For example if we add another set that happens to also have dice icons.

Gazook89 reviewed (changes requested):
Added a couple of tiny notes, and added a PR to your branch to include tests.

calculuschild commented on shared/naturalcrit/codeEditor/autocomplete-emoji.js:
Good question. My first thought is we should try to at least follow what Github does.

```
:heart::star:
```

renders as

:heart::star:

But we can think a little about this.

calculuschild replied:
Thinking about this, I'm worried we may be stuck unless we change the definition list.  That space does affect the render output, and some people might not want the space there.

![image](https://github.com/naturalcrit/homebrewery/assets/5878534/408ba689-3608-46b1-bace-a571c99bb456)

On the other hand, if we say definition lists *must* have a space before and after the `::` then that would resolve this, since `Definition::Term` renders the same as `Definition :: term`. But I don't know how disruptive that would be to existing brews.

5e-Cleric replied:
I think we should not disrupt dl at all, even if some people don't want those spaces, i think that is the best way to solve this

Gazook89 replied:
I’m right now of the opposite opinion: down with DLs!

G-Ambatte replied:
I think that we should NOT disrupt DLs if we can avoid it; they've been seized upon and used extensively by many authors. The hanging indentation of the DL is also useful for Pathfinder authors, as it's a common styling in the Paizo source books.

That said, I'm not sure why the text that is parsed as an emoji block is still considered a DL - as I understand it, the text considered the source of the emoji should be removed from the raw text, leaving no `:` to be parsed as a DL.

calculuschild replied:
> The hanging indentation of the DL is also useful for Pathfinder authors, as it's a common styling in the Paizo source books.

This would still continue function. We would just be enforcing spacing around the `::`. The output would be the same.

So `This wouldn't::work` but `This would :: work` fine.

But I see the point that it is probably used more often than it seems at first glance, which means more chance of disruption.

> That said, I'm not sure why the text that is parsed as an emoji block is still considered a DL - as I understand it, the text considered the source of the emoji should be removed from the raw text, leaving no `::` to be parsed as a DL.

Discussed in the Gitter, but for others: Block-level tokens are processed first, then their text contents are processed for inline-level tokens. Definition list can't start partway through a line. It contains the entire line, thus it is a block token. Emojis inside the line aren't processed until the DL has already been split into DD and DT.

calculuschild replied:
> Further to this, the autocomplete currently detects entering multiple : to create a vertical space is detected as the user starting an emoji.

This has been fixed now. Autocomplete should only trigger on non-space/non-colon characters after the initial `:`

calculuschild commented on themes/fonts/icon fonts/dicefont.js:
Alphabetical makes sense. We should be able to just apply a sorting filter inside the auto complete so we don't have to manually sort each list.

calculuschild replied:
Fixed. Also sorts in a more logical numerical sequence now, so

d2, d4, d6, d8, d10, d12, d20, instead of d10, d12, d2, d20, d4, d6, d8

So all d2's stay together and d20's stay together, instead of alternating.

calculuschild commented on shared/naturalcrit/codeEditor/codeEditor.less:
```diff
@@ -2,6 +2,8 @@
 @import (less) 'codemirror/addon/search/matchesonscrollbar.css';
 @import (less) 'codemirror/addon/dialog/dialog.css';
+@import (less) 'codemirror/addon/hint/show-hint.css';
+@import (less) './themes/fonts/icon fonts/dicefont.less';
```
I would prefer to keep this PR touching only the essential changes.

calculuschild commented on shared/naturalcrit/codeEditor/codeEditor.jsx:
```diff
@@ -36,6 +37,8 @@ if(typeof window !== 'undefined'){
 	//XML code folding is a requirement of the auto-closing tag feature and is not enabled
 	require('codemirror/addon/fold/xml-fold.js');
 	require('codemirror/addon/edit/closetag.js');
+	//Autoccompletion
```
fixed

calculuschild commented:
For reference, changing the DL syntax would just be this adjustment to the regex, adding two spaces:

![image](https://github.com/naturalcrit/homebrewery/assets/5878534/cf7f334f-43e6-451e-a910-13181e855b12)

calculuschild commented:
I hacked together a modification to the DefinitionList extension to specifically check for emojis inside, and ignore them for the purposes of detecting a `::`. This seems to resolve the conflicts as far as I can see, but in V4 we should remove the hack and just update the DefinitionList regex to require spaces.

calculuschild commented on shared/naturalcrit/markdown.js:
```diff
@@ -304,10 +306,19 @@ const definitionListsSingleLine = {
 		let endIndex = 0;
 		const definitions = [];
 		while (match = regex.exec(src)) {
```
I recognize it is a little excessive to run this regex here where the `match` result isn't used anymore (except to grab the full line). But I'm hoping to leave this extension mostly intact until V4 comes around where we can do a proper cleanup.

5e-Cleric replied:
I think keeping it clean now will help us later on, we might not remember to do this or we might have other PRs in mind. If it's not too much to ask.

calculuschild commented:
At this point, I just want to add fontAwesome to the autocomplete, and settle on a naming convention for the emojis (is camelCase ok? UpperCamelCase?

`:EiBarbarianRage:` vs `:ei-barbarian-rage:`, etc.

Gazook89 commented:
Just because I did the work and need a place to save it, here is a list of links for all of the default roll20 token marker icons on game-icons.net.

I was planning to put the image in the r20 column, but realized I didn't care that much.  Here is a screenshot of the batch of them-- the table shows them in order from left to right, skipping the first few boring ones.

<img width="256" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/6d7ea819-64e8-43ef-a673-d1cf50d25d3b">

Note, that at least on my r20, that it seems r20 has a slightly modified color scheme from the default on game-icons.  game-icons lets you modify the svg's right on their site, or it oculd be done with another svg editor.  Maybe r20's icons change if you change the website dark mode or not.

| r20 | game-icons.net                                                                     |
| --- | ---------------------------------------------------------------------------------- |
|     | [gi_skull_crossed_bones](https://game-icons.net/1x1/lorc/skull-crossed-bones.html) |
|     | [gi_sleepy](https://game-icons.net/1x1/lorc/sleepy.html)                           |
|     | [gi_half_heart](https://game-icons.net/1x1/lorc/half-heart.html)                   |
|     | [gi_heat_haze](https://game-icons.net/1x1/lorc/heat-haze.html)                     |
|     | [gi_interdiction](https://game-icons.net/1x1/lorc/interdiction.html)               |
|     | [gi_snail](https://game-icons.net/1x1/lorc/snail.html)                             |
|     | [gi_lightning_helix](https://game-icons.net/1x1/lorc/lightning-helix.html)         |
|     | [gi_spanner](https://game-icons.net/1x1/lorc/spanner.html)                         |
|     | [gi_chained_heart](https://game-icons.net/1x1/lorc/chained-heart.html)             |
|     | [gi_chemical_bolt](https://game-icons.net/1x1/lorc/chemical-bolt.html)             |
|     | [gi_death_zone](https://game-icons.net/1x1/lorc/death-zone.html)                   |
|     | [gi_drink_me](https://game-icons.net/1x1/lorc/drink-me.html)                       |
|     | [gi_edge_crack](https://game-icons.net/1x1/lorc/edge-crack.html)                   |
|     | [gi_ninja_mask](https://game-icons.net/1x1/lorc/ninja-mask.html)                   |
|     | [gi_stopwatch](https://game-icons.net/1x1/lorc/stopwatch.html)                     |
|     | [gi_fishing_net](https://game-icons.net/1x1/lorc/fishing-net.html)                 |
|     | [gi_overdrive](https://game-icons.net/1x1/lorc/overdrive.html)                     |
|     | [gi_strong](https://game-icons.net/1x1/lorc/strong.html)                           |
|     | [gi_fist](https://game-icons.net/1x1/lorc/fist.html)                               |
|     | [gi_padlock](https://game-icons.net/1x1/lorc/padlock.html)                         |
|     | [gi_three_leaves](https://game-icons.net/1x1/lorc/three-leaves.html)               |
|     | [gi_fluffy_wing](https://game-icons.net/1x1/lorc/fluffy-wing.html)                 |
|     | [gi_pummeled](https://game-icons.net/1x1/lorc/pummeled.html)                       |
|     | [gi_tread](https://game-icons.net/1x1/lorc/tread.html)                             |
|     | [gi_arrowed](https://game-icons.net/1x1/lorc/arrowed.html)                         |
|     | [gi_aura](https://game-icons.net/1x1/lorc/aura.html)                               |
|     | [gi_back_pain](https://game-icons.net/1x1/lorc/back-pain.html)                     |
|     | [gi_black_flag](https://game-icons.net/1x1/lorc/black-flag.html)                   |
|     | [gi_bleeding_eye](https://game-icons.net/1x1/lorc/bleeding-eye.html)               |
|     | [gi_bolt_shield](https://game-icons.net/1x1/lorc/bolt-shield.html)                 |
|     | [gi_broken_heart](https://game-icons.net/1x1/lorc/broken-heart.html)               |
|     | [gi_cobweb](https://game-icons.net/1x1/lorc/cobweb.html)                           |
|     | [gi_broken_shield](https://game-icons.net/1x1/lorc/broken-shield.html)             |
|     | [gi_flying_flag](https://game-icons.net/1x1/lorc/flying-flag.html)                 |
|     | [gi_radioactive](https://game-icons.net/1x1/lorc/radioactive.html)                 |
|     | [gi_trophy](https://game-icons.net/1x1/lorc/trophy.html)                           |
|     | [gi_broken_skull](https://game-icons.net/1x1/lorc/broken-skull.html)               |
|     | [gi_frozen_orb](https://game-icons.net/1x1/lorc/frozen-orb.html)                   |
|     | [gi_rolling_bomb](https://game-icons.net/1x1/lorc/rolling-bomb.html)               |
|     | [gi_white_tower](https://game-icons.net/1x1/lorc/white-tower.html)                 |
|     | [gi_grab](https://game-icons.net/1x1/lorc/grab.html)                               |
|     | [gi_screaming](https://game-icons.net/1x1/lorc/screaming.html)                     |
|     | [gi_grenade](https://game-icons.net/1x1/lorc/grenade.html)                         |
|     | [gi_sentry_gun](https://game-icons.net/1x1/lorc/sentry-gun.html)                   |
|     | [gi_all_for_one](https://game-icons.net/1x1/lorc/all-for-one.html)                 |
|     | [gi_angel_outfit](https://game-icons.net/1x1/lorc/angel-outfit.html)               |
|     | [gi_archery_target](https://game-icons.net/1x1/lorc/archery-target.html)           |

Gazook89 commented:
I thought I saw a notification somewhere that you had another minor PR or something that took care of the directory renaming here but couldn't spot it right away...

It might be worth throwing each icon font into it's own filter, before this gets messy:
```
icon fonts/
┣ diceFont.js
┣ diceFont.less
┣ diceFont.woff2
┣ diceFont_license.md
┣ elderberryInn.js
┣ elderberryInn.less
┗ elderberryInn.woff2
```

5e-Cleric commented:
## Any idea as to what causes this highlight? this is correct syntax.

![image](https://github.com/naturalcrit/homebrewery/assets/91662974/433e8323-1540-4316-a425-5c703e4d09e6)

5e-Cleric commented:
Can we allow inline css inject? `My icon :df-d20-20:{font-size:30px}
`
If only to avoid unnecessary spans

5e-Cleric commented:
Unsure if we should prevent this, maybe check if this counts as css first.

![image](https://github.com/naturalcrit/homebrewery/assets/91662974/84943912-7b5c-4c15-bb31-8aec651148a1)

Gazook89 commented:
Regarding

> Unsure if we should prevent this, maybe check if this counts as css first

I don’t think that is to much bother to have the menu pop— as you continue typing it goes away and to me it’s just reminiscent of autocomplete in places like vscode

calculuschild commented:
> Unsure if we should prevent this, maybe check if this counts as css first.
> 
> ![image](https://private-user-images.githubusercontent.com/91662974/324243844-84943912-7b5c-4c15-bb31-8aec651148a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzNTc5MTYsIm5iZiI6MTcxNDM1NzYxNiwicGF0aCI6Ii85MTY2Mjk3NC8zMjQyNDM4NDQtODQ5NDM5MTItN2I1Yy00YzE1LWJiMzEtOGFlYzY1MTE0OGExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI5VDAyMjY1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzOWI1OTc4ZGQ2OGU1YTFkNWFkOTE2MDFhZmNkMDZmY2MzY2UyMzQxYWFlOTAwMjA3NzIwNWUzN2IwNjhmMzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I1uIWWV1D1tY9OT3LctkXC9uUXRwt3iiZKhTHX83aAc)

Pushed an update that should fix this now. Still working on the other items.

Gazook89 commented:
@calculuschild did you want to add the tests from #3406?

calculuschild commented:
> @calculuschild did you want to add the tests from #3406?

Yeah. I'm going to finish addressing the issues from 5e-Cleric first in case that changes how we want our tests to work, then I will copy those over.

Gazook89 commented:
With your last commit, in the deployment, it crashes to a white screen (in FF and Chrome):

```
bundle.js:35609 Uncaught ReferenceError: match is not defined
    at bundle.js:79715:17
    at arrayEach (bundle.js:55470:11)
    at Function.forEach (bundle.js:64350:14)
    at bundle.js:79546:11
    at runInOp (bundle.js:48993:18)
    at CodeMirror.operation (bundle.js:53701:37)
    at ReactClassComponent.highlightCustomMarkdown (bundle.js:79537:18)
    at ReactClassComponent.componentDidMount (bundle.js:79484:10)
    at ReactClassComponent.chainedFunction [as componentDidMount] (bundle.js:2727:11)
    at lk (bundle.js:35743:217)
```

calculuschild commented:
> With your last commit, in the deployment, it crashes to a white screen (in FF and Chrome):

Missed committing one file. Should work now.

Gazook89 commented:
just going to drop a screencap of what this deployment looks like, in terms of highlight styling:

<img width="299" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/6c78c4da-de3c-47b7-b23e-4448fb60a8ab">

------
## definition lists styling
One issue I see that could be improved is that it breaks definition list highlighting:
<img width="743" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/0f73d8ed-cce7-4311-aa27-f333b4ca4d77">

The above should have a highlight on both DL's, but only displays it on the second.  The first one in fact does have the right classes for the `definition` and `::`, but not for the `term`, as shown in the HTML:

```html
<span role="presentation">
	<span class="cm-builtin emoji">:df-d8-1:</span>
	<span class=" dd-highlight dl-colon-highlight">::</span>
	<span class=" dd-highlight"> Consectetur non laborum aliquip. Deserunt in veniam do sunt id ullamco aliqua non pariatur et. Culpa eiusmod dolor aliqua est magna enim. Veniam aliqua esse et. Id deserunt eu fugiat eiusmod magna pariatur adipisicing incididunt tempor nulla cillum exercitation. Et veniam proident laborum incididunt elit adipisicing reprehenderit occaecat amet eu.</span>
</span>
```
I think we still would want the emoji highlight style to be shown over the term, but it shouldn't prevent the rest of the defintion list highlighting.

For reference, here is the DL just below it, with correct display:
```html
<span role="presentation">
	<span class=" dl-highlight dt-highlight">Officia </span>
	<span class=" dl-highlight dd-highlight dl-colon-highlight">::</span>
	<span class=" dl-highlight dd-highlight"> consequat labore cillum mollit consectetur labore anim occaecat eu. Ipsum id ea voluptate deserunt. Et officia consequat eu tempor. Nostrud ex sint adipisicing qui qui minim minim labore ad et.</span>
</span>
```

## header styling
Here is how it looks with an emoji in a header:
<img width="557" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/29a4bb1d-f589-4117-82ea-25f05358763f">

It may be that headers are very important for navigating the Editor pane-- being able to find the titles of sections while scrolling through a large block of monospace text is crucial.  So I think that headers should have a visual consistency, which is foiled by having the emoji styled so differently.  I think I'll have general suggestions on the styling at the end of this PR, but at minimum for headers I think that the default header styling should win out.

------

Some thoughts on alternative styling:

1. Only use `background-color`, and somehow set it to be a variant of the existing text color. I haven't really thought about how to do this, but throwing it out there.  Something like this:
<img width="557" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/b163dbd5-fe59-453e-838f-416cff41bbc9">

2. Maybe just toning it down a bit, throwing in some italics?
<img width="557" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/c337f5e1-be55-4c40-9d2e-b0710bede097">

3. Do even more JS and do what other editors do:  turn the icon syntax directly into the rendered icon within the editor.  Treat it still as the full syntax text, though, so that when you hit delete button it reverts to the text, and when you copy the icon it pastes as the full text.

I think that this emoji syntax, and the definition list `::` are the only two syntax that use a background color for the highlight.  I don't think a long discussion about whether this is a good thing or bad thing is necessary, but mentioning it in case it inspires some thought.

Finally, how does this all work with various editor themes?  Who knows.

5e-Cleric commented:
I'd actually prefer the pink highlight, more contrast makes it easier to see, but can we see this with dark themes?

5e-Cleric commented:
If you don't mind, could you add in a comment here a little steps on how to add an icon font in the future to the emoji syntax? a simple "add this in that file and that in this other file" will suffice. And it'd make things easier in the long run.

calculuschild commented:
> I'd actually prefer the pink highlight, more contrast makes it easier to see, but can we see this with dark themes?

Dark themes can be edited with their own coloring in `themes/codeMirror/customEditorStyles.less`. I'm going to leave that exercise to the reader in another PR though for someone who is better at graphical design.

> If you don't mind, could you add in a comment here a little steps on how to add an icon font in the future to the emoji syntax? a simple "add this in that file and that in this other file" will suffice. And it'd make things easier in the long run.

Yeah I will put a little comment into `markdown.js`

calculuschild commented:
Merging! Let the chaos ensue.

5e-Cleric commented:
> Merging! Let the chaos ensue.

This message does not reflect the intentions of the Homebrewery Dev team and does in no way give permission to anyone or anything to display chaotic behaviour in any specially disrupting way.

This joke commment was brought to you by the meme layer association.
